### PR TITLE
llms/googleai: Support more `option.ClientOption` and update google ai sdk

### PR DESCRIPTION
**Pull request by wangjiancn**
Support `github.com/google/generative-ai-go/genai` `option.ClientOption`:

- WithCredentialsJSON
- WithCredentialsFile
- WithHttpClient

> Without WithCredentialsFile, vertex ai cannot run in non-GCP environments.

Fixe: #838

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of the primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `Fixes #123`).
- [x] Describes the source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

**Comment by skrishnan-sap** [comment]
@wangjiancn Thank you for this PR. I was also adding `WithHTTPClient` option with #840 . I will close my PR, but would like to ask you whether you would like to take over the unit test added for the scenario?

**Comment by wangjiancn** [comment]
> @wangjiancn Thank you for this PR. I was also adding `WithHTTPClient` option with #840 . I will close my PR, but would ask whether you would like to take over the unit test I added for the scenario from my PR?

Sure, I'll take a moment to incorporate the test case. I've just been handling the configuration, and I understand it should work without issues.

**Comment by wangjiancn** [comment]
@skrishnan-sap 

```
	github.com/tmc/langchaingo v0.1.10 => github.com/wangjiancn/langchaingo v0.1.11-dev3
```

File: llms/googleai/palm/palm_llm.go [hunk 1]
```diff
@@ -111,6 +112,9 @@ func newClient(opts ...Option) (*palmclient.PaLMClient, error) {
 	if len(options.projectID) == 0 {
 		return nil, ErrMissingProjectID
 	}
+	if len(options.location) == 0 {
+		return nil, ErrMissingLocation
+	}
 
-	return palmclient.New(options.projectID, options.clientOptions...)
+	return palmclient.New(context.TODO(), options.projectID, options.location, options.clientOptions...)
```
Comment (eliben) [inline llms/googleai/palm/palm_llm.go:119]: There was a default location before - has it been removed?

File: examples/vertex-completion-example/vertex-completion-example.go [hunk 2]
```diff
@@ -18,7 +20,13 @@ func main() {
 	ctx := context.Background()
 	project := os.Getenv("VERTEX_PROJECT")
 	location := os.Getenv("VERTEX_LOCATION")
-	llm, err := vertex.New(ctx, googleai.WithCloudProject(project), googleai.WithCloudLocation(location))
+	credentialsJSONFile := os.Getenv("VERTEX_CREDENTIALS")
```
Comment (eliben) [inline examples/vertex-completion-example/vertex-completion-example.go:23]: So this example now requires a JSON credentials file?

File: llms/googleai/internal/palmclient/palmclient.go [hunk 3]
```diff
@@ -48,19 +42,20 @@ type PaLMClient struct {
...
-func New(projectID string, opts ...option.ClientOption) (*PaLMClient, error) {
+func New(ctx context.Context, projectID, location string, opts ...option.ClientOption) (*PaLMClient, error) {
 	numConns := runtime.GOMAXPROCS(0)
 	if numConns > defaultMaxConns {
 		numConns = defaultMaxConns
 	}
 	o := []option.ClientOption{
 		option.WithGRPCConnectionPool(numConns),
-		option.WithEndpoint(defaultAPIEndpoint),
+		option.WithEndpoint(fmt.Sprintf("%s-aiplatform.googleapis.com:443", location)),
```
Comment (wangjiancn) [inline llms/googleai/internal/palmclient/palmclient.go:52]: @eliben I felt the previous endpoint is hard code and couldn't be set, possibly due to a bug, so I modified it.

File: llms/googleai/palm/palm_llm.go (hunk 1 above)
Comment (wangjiancn) [reply llms/googleai/palm/palm_llm.go:119]: I felt the previous endpoint was dead and couldn't be set, possibly due to a bug, so I modified it.

File: llms/googleai/internal/palmclient/palmclient.go (hunk 3 above)
Comment (eliben) [reply llms/googleai/internal/palmclient/palmclient.go:52]: I suggest making fewer changes at once, to make the PR easier to review?
Can you please adjust the PR to just add the new option? You can sync with head - it should have the latest go.mod/go.sum changes, so please don't include those in the PR either

Comment (wangjiancn) [reply llms/googleai/internal/palmclient/palmclient.go:52]: Understood. I have already been using this branch in the production environment. I will make the necessary adjustments when I find some time.

**Comment by xavidop** [comment]
looking forward to this PR!

**Comment by tmc** [comment]
@wangjiancn would love to see this get over the line!

**Comment by wangjiancn** [comment]
@tmc I've updated the PR, could you please review it? The related code has been used in the production environment, so please help to merge it first.

**Comment by eliben** [comment]
Just a quick note: I'll be happy to review this PR, but am on vacation now and will be back in about a week. Hopefully it can wait until then

**Review by xavidop — APPROVED** [review]

**Comment by xavidop** [comment]
It looks good to me!

File: examples/vertex-completion-example/go.mod [hunk 4]
```diff
@@ -7,44 +7,47 @@ toolchain go1.22.1
...
+	google.golang.org/api v0.180.0 // indirect
+	google.golang.org/genproto v0.0.0-20240401170217-c3f982113cda // indirect
+	google.golang.org/genproto/googleapis/api v0.0.0-20240515191416-fc5f0ca64291 // indirect
+	google.golang.org/genproto/googleapis/rpc v0.0.0-20240515191416-fc5f0ca64291 // indirect
+	google.golang.org/grpc v1.64.0 // indirect
+	google.golang.org/protobuf v1.34.1 // indirect
 )
+
+// test for new pkg version
+replace github.com/tmc/langchaingo => ../..
```
Comment (tmc) [inline examples/vertex-completion-example/go.mod:53]: Generally we prefer use of go.work files over replace directives.

**Review by tmc — APPROVED** [review]
While I agree with Eli's comments regarding scope, I think this should make it in. LGTM.

**Pull request merged**